Forests per host

dmcassel opened the pull request:
Implemented the forests-per-host setting. If PR is accepted, I'll write a wiki showing how to use it.

dmcassel committed:
Reads forests-per-host property and uses it; still need to handle wipe

dmcassel committed:
#5 revise forest removal to detach and remove the current forests
Previously removed forests based on what the configuration says, but
it's more reliable to get rid of what's attached.

dmcassel committed:
#5 improving error messages

dmcassel committed:
#5 message cleanup

grtjn commented on src/lib/modules/databases.js:
```diff
@@ -29,11 +29,62 @@ DBManager.databaseOperation = function(operation, database, callback) {
+        var hostSettings = common.objectSettings('hosts/' + host, this.env);
+        console.log('host is: ' + hostSettings['host-name']);
+        for (var i = 0; i < count; i++) {
+            forests.push({'host': hostSettings['host-name'], 'forest-name': dbSettings['database-name'] + '-' + (index * count + i)});
```
While at this, you could apply the same naming convention that Roxy applies: dbname-hostnr-forestnr, where hostnr is zero-padded to 3 digits.. :)

grtjn commented on src/bin/mlsound-wipe.js:
```diff
@@ -36,12 +36,14 @@ prompt.get({
-                                dbManager.removeRoles(function() {
-                                    logger.info('Project wiped!');
+                dbManager.removeForests('content', 'full', function() {
+                    dbManager.removeForests('modules', 'full', function() {
```
Does this work? Shouldn't you detach forests first, then delete the databases, then delete the forests? (that is how Roxy does it ;) Or maybe detach+delete forests, then delete databases, but detach of forests must come first..

dmcassel replied:
The removeForests function detaches and deletes the forests.

grtjn replied:
Yes, I saw that, but you are calling removeDatabase first (you switched that around for some reason)..

dmcassel replied:
No, I'm calling removeForests first. mlsound uses a callback mechanism, so this removes content forests, then modules forests, then the content database, etc.

grtjn replied:
Ah, overlooked that. Writing it as a chain would make it easier to read.. :-P

dmcassel replied:
My preference would be using Promises, but I went with the project's style.

miguelrgonzalez replied:
I was debating which of the approaches to pick, and ended up with callbacks...
To be completely honest I'm not 100% happy with it neither, and chances are I will eventually move it to promises.

dmcassel commented:
Don't accept this PR yet; I sometimes see a timing problem (attempting database creation before MarkLogic recognizes that the forests are ready). Thought I'd cleared that up, but I'll give it another look.

rjrudin commented:
Food for thought - in ml-app-deployer, I went with a command approach, where each command is responsible for some portion of the deployment - e.g. a command for creating databases, a command for creating users, etc. Each command defines an order that it executes in during a deploy and during an undeploy (e.g. during deploy, a database has to be created before a server, and during an undeploy, a server has to be deleted before the database is deleted). 

The deployer object is then unaware of what any command does - he's just given a list of commands to execute, and he sorts them (based on whether he's doing a deploy or undeploy), and then executes them. This gives a nice open/closed effect where the deployer can be modified by adding new commands, and the order can be precisely controlled, all without modifying the deployer. 

I think the equivalent here would be that instead of wipe.js having explicit knowledge of everything that must be done and in what order, it would instead load up a list of commands (or have them injected somehow), sort them, and then execute them. Just something to consider.

dmcassel committed:
#5 revised to properly handle async requests

dmcassel commented:
Updated code to properly handle asynchronous checking for and creation of forests. @miguelrgonzalez, how does this look to you?

miguelrgonzalez commented:
Hi,
thanks for the pull!!

It does look to me, only a couple of things:
- I see some console.log, please use the logger instead. There is some "nice" formatting around the output.
- Instead of using 

```
    --outStandingRequests; 
    checkFinished();
```

consider using callBackwhenDone() like the rest of the methods. It's functionally the same, but just for the sake of consistency.
- When splitting functionality in modules, I thought that it made sense to have forest operations under servers.js (that's were the remove operations are), but I see you placed buildForestsByHost function under databases. This function should be moved to keep things consistent.

Again, thanks a lot for expending time on this!
Cheers,
Miguel

dmcassel committed:
#5 code cleanup
- removed console.log or changed to logger
- switched to using callBackwhenDone
- moved buildForestsByHost to servers.js

dmcassel commented:
@miguelrgonzalez, updated as specified.

miguelrgonzalez commented:
@dmcassel Looks like accepting your other 2 pull requests, created some conflicts...

dmcassel commented:
@miguelrgonzalez Conflicts fixed

miguelrgonzalez commented:
merged!
